File: 18powernumber2/power_number.cpp
#include <iostream>
using namespace std;

double pow(int a, int n){
  double res = 1;
  while(n > 0){
    if(n & 1){
      // 1bit目がたっていればaをかける
      res *= a;
    }
    //
    a = a*a;
    // bitをずらす
    // 1010->101->10->1
    n >>= 1;
  }
  return res;
}

int main(int argc, char const *argv[])
{
  cout << pow(2,0) << endl;
  cout << pow(2, 1) << endl;
  cout << pow(2, 2) << endl;
  cout << pow(2,10) << endl;
  cout << pow(3,3) << endl;
  //cout << pow(2, 1024) << endl;
  //cout << pow(2, 10000007) << endl;

  return 0;
}

File: 18powernumber2/power_number.py
def pow(a, n):
  res = 1
  while n > 0:
    if n & 1: res = res * a
    a **= 2
    n >>= 1
  return res

print(pow(2, 0))
print(pow(2, 1))
print(pow(2, 2))
print(pow(2, 10))
print(pow(2, 1024))
# print(pow(2, 1000007)) # だいたい10秒ぐらい。
# print(2 ** 1000007)  # これも10秒ぐらいだった。
